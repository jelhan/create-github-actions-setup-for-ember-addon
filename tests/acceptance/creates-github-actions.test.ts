import execa from 'execa';
import fs from 'fs';
import { mkdtemp, readFile, rmdir } from 'fs/promises';
import { copy } from 'fs-extra';
import os from 'os';
import path from 'path';

const executable = path.join(
  __dirname,
  '..',
  '..',
  'bin',
  'create-github-actions-setup-for-ember-addon'
);

let tmpDirForTesting: string;

beforeEach(async () => {
  tmpDirForTesting = await mkdtemp(
    path.join(os.tmpdir(), 'create-github-actions-setup-for-ember-addon-tests-')
  );
});
afterEach(async () => {
  await rmdir(tmpDirForTesting, { recursive: true });
});

describe('creates GitHub Actions setup', () => {
  it('uses default values if no TravisCI configuration exists', async () => {
    await execa(executable, [], {
      cwd: tmpDirForTesting,
    });

    expect(
      await readFile(
        path.join(tmpDirForTesting, '.github', 'workflows', 'ci.yml'),
        { encoding: 'utf-8' }
      )
    ).toMatchSnapshot();
  });

  describe('picks up configuration from previous run stored in GitHub Actions workflow', () => {
    const fixturesPath = path.join(
      __dirname,
      '..',
      'fixtures',
      'github-actions'
    );
    const scenarios = fs.readdirSync(fixturesPath);

    scenarios.forEach((scenario) => {
      it(`supports scenario ci.yml.${scenario}`, async () => {
        await copy(path.join(fixturesPath, scenario), tmpDirForTesting);

        await execa(executable, [], {
          cwd: tmpDirForTesting,
        });

        expect(
          await readFile(
            path.join(tmpDirForTesting, '.github', 'workflows', 'ci.yml'),
            { encoding: 'utf-8' }
          )
        ).toMatchSnapshot();
      });
    });
  });

  describe('migrates existing TravisCI configration', () => {
    const fixturesPath = path.join(__dirname, '..', 'fixtures', 'travis-ci');
    const scenarios = fs.readdirSync(fixturesPath);

    scenarios.forEach((scenario) => {
      it(`supports scenario .travis.yml.${scenario}`, async () => {
        await copy(path.join(fixturesPath, scenario), tmpDirForTesting);

        await execa(executable, [], {
          cwd: tmpDirForTesting,
        });

        expect(
          await readFile(
            path.join(tmpDirForTesting, '.github', 'workflows', 'ci.yml'),
            { encoding: 'utf-8' }
          )
        ).toMatchSnapshot();
      });
    });
  });
});
