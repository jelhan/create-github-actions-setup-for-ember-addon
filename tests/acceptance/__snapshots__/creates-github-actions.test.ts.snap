// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates GitHub Actions setup migrates existing TravisCI configration supports scenario .travis-ci.yml.ember-3.20-npm 1`] = `
"name: CI

on: [push, pull_request]

jobs:
  lint:
    name: Lint
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: npm ci
    - name: Lint
      run: npm run-script lint

  test:
    name: Tests
    runs-on: \${{ matrix.os }}
    needs: lint

    strategy:
      matrix:
        os: [ubuntu-latest]
        browser: [chrome]

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: npm ci
    - name: Test
      run: npm run-script test:ember --launch \${{ matrix.browser }}

  floating-dependencies:
    name: Floating Dependencies
    runs-on: \${{ matrix.os }}
    needs: lint

    strategy:
      matrix:
        os: [ubuntu-latest]
        browser: [chrome]

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: npm install --no-package-lock
    - name: Test
      run: npm run-script test:ember --launch \${{ matrix.browser }}

  try-scenarios:
    name: Tests - \${{ matrix.ember-try-scenario }}
    runs-on: ubuntu-latest
    continue-on-error: \${{ matrix.allow-failure }}
    needs: test

    strategy:
      fail-fast: true
      matrix:
        ember-try-scenario: [
          ember-lts-3.12,
          ember-lts-3.16,
          ember-release,
          ember-beta,
          ember-default-with-jquery,
          ember-classic
        ]
        allow-failure: [false]
        include:
          - ember-try-scenario: ember-canary
            allow-failure: true

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: npm ci
    - name: Test
      env:
        EMBER_TRY_SCENARIO: \${{ matrix.ember-try-scenario }}
      run: node_modules/.bin/ember try:one $EMBER_TRY_SCENARIO
"
`;

exports[`creates GitHub Actions setup migrates existing TravisCI configration supports scenario .travis-ci.yml.ember-3.20-yarn 1`] = `
"name: CI

on: [push, pull_request]

jobs:
  lint:
    name: Lint
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: yarn install --frozen-lockfile
    - name: Lint
      run: yarn lint

  test:
    name: Tests
    runs-on: \${{ matrix.os }}
    needs: lint

    strategy:
      matrix:
        os: [ubuntu-latest]
        browser: [chrome]

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: yarn install --frozen-lockfile
    - name: Test
      run: yarn test:ember --launch \${{ matrix.browser }}

  floating-dependencies:
    name: Floating Dependencies
    runs-on: \${{ matrix.os }}
    needs: lint

    strategy:
      matrix:
        os: [ubuntu-latest]
        browser: [chrome]

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: yarn install --no-lockfile --non-interactive
    - name: Test
      run: yarn test:ember --launch \${{ matrix.browser }}

  try-scenarios:
    name: Tests - \${{ matrix.ember-try-scenario }}
    runs-on: ubuntu-latest
    continue-on-error: \${{ matrix.allow-failure }}
    needs: test

    strategy:
      fail-fast: true
      matrix:
        ember-try-scenario: [
          ember-lts-3.12,
          ember-lts-3.16,
          ember-lts-3.20,
          ember-release,
          ember-beta,
          ember-default-with-jquery,
          ember-classic
        ]
        allow-failure: [false]
        include:
          - ember-try-scenario: ember-canary
            allow-failure: true

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: yarn install --frozen-lockfile
    - name: Test
      env:
        EMBER_TRY_SCENARIO: \${{ matrix.ember-try-scenario }}
      run: node_modules/.bin/ember try:one $EMBER_TRY_SCENARIO
"
`;

exports[`creates GitHub Actions setup uses default values if no TravisCI configuration exists 1`] = `
"name: CI

on: [push, pull_request]

jobs:
  lint:
    name: Lint
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: yarn install --frozen-lockfile
    - name: Lint
      run: yarn lint

  test:
    name: Tests
    runs-on: \${{ matrix.os }}
    needs: lint

    strategy:
      matrix:
        os: [ubuntu-latest]
        browser: [chrome, firefox]

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: yarn install --frozen-lockfile
    - name: Test
      run: yarn test:ember --launch \${{ matrix.browser }}

  floating-dependencies:
    name: Floating Dependencies
    runs-on: \${{ matrix.os }}
    needs: lint

    strategy:
      matrix:
        os: [ubuntu-latest]
        browser: [chrome, firefox]

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: yarn install --no-lockfile --non-interactive
    - name: Test
      run: yarn test:ember --launch \${{ matrix.browser }}

  try-scenarios:
    name: Tests - \${{ matrix.ember-try-scenario }}
    runs-on: ubuntu-latest
    continue-on-error: \${{ matrix.allow-failure }}
    needs: test

    strategy:
      fail-fast: true
      matrix:
        ember-try-scenario: [
          ember-lts-3.16,
          ember-lts-3.20,
          ember-release,
          ember-beta,
          ember-default-with-jquery,
          ember-classic
        ]
        allow-failure: [false]
        include:
          - ember-try-scenario: ember-canary
            allow-failure: true
          - ember-try-scenario: embroider-tests
            allow-failure: true

    steps:
    - uses: actions/checkout@v2
      with:
        fetch-depth: 1
    - uses: actions/setup-node@v2-beta
      with:
        node-version: '10.x'
    - name: Install Dependencies
      run: yarn install --frozen-lockfile
    - name: Test
      env:
        EMBER_TRY_SCENARIO: \${{ matrix.ember-try-scenario }}
      run: node_modules/.bin/ember try:one $EMBER_TRY_SCENARIO
"
`;
